staging/src/github.com/openshift/openshift-apiserver/pkg/project/auth/cache.go

```diff
@@ -246,16 +250,37 @@ func NewAuthorizationCache(
 	}
 	ac.lastSyncResourceVersioner = namespaceLastSyncResourceVersioner
 	ac.syncHandler = ac.syncRequest
+
+	ac.informerCachesToSync = append(ac.informerCachesToSync, []cache.InformerSynced{
+		informers.RoleBindings().Informer().HasSynced,
+		informers.Roles().Informer().HasSynced,
+		informers.ClusterRoleBindings().Informer().HasSynced,
+		informers.ClusterRoles().Informer().HasSynced,
+	}...)
+
 	return ac
 }
 
 // Run begins watching and synchronizing the cache
-func (ac *AuthorizationCache) Run(period time.Duration) {
+func (ac *AuthorizationCache) Run(stopCh <-chan struct{}, period time.Duration) {
 	ac.skip = &statelessSkipSynchronizer{}
 
+	if !cache.WaitForCacheSync(stopCh, ac.informerCachesToSync...) {
+		utilruntime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+	}
+	ac.informerCachesSyncedLock.Lock()
+	ac.informerCachesSynced = true
+	ac.informerCachesSyncedLock.Unlock()
+
 	go utilwait.Forever(func() { ac.synchronize() }, period)
 }
 
+func (ac *AuthorizationCache) hasCachesSynced() bool {
+	ac.informerCachesSyncedLock.Lock()
```

Comment:
use a RW lock. This is in the critical path.

Reply:
sure, fixed.